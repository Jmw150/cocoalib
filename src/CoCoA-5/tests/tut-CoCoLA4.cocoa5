------------------------------------------------------------
-- TUTORIAL 4
------------------------------------------------------------
-- SOLVING ZERO DIMENSIONAL SYSTEMS
------------------------------------------------------------

-- ring for 0-dimensional ideal
P ::= QQ[x,y,z];

------------------------------------------------------------
-- RATIONAL SOLUTIONS
------------------------------------------------------------
use P;

I := IntersectList([ideal(z^2-4, y^3-8, (z-x)*(x-3)),
		    ideal(x^2-1, y^2-1, x+y+z)
		    ]);
indent(I);

--------------------------------------------------
-- THE "BAD" ALGORITHM (brute force)

EFx := eigenfactors(MultiplicationMat(x, I),  x);
Xs := [NR(x,[f]) | f in EFx and deg(f)=1 ];  Xs;

EFy := eigenfactors(MultiplicationMat(y, I),  y);
Ys := [NR(y,[f]) | f in EFy and deg(f)=1 ];  Ys;

EFz := eigenfactors(MultiplicationMat(z, I),  z);
Zs := [NR(z,[f]) | f in EFz and deg(f)=1 ];  Zs;

define IsSolution(XYZ, I)
  foreach g in gens(I) do
    if not(IsZero(eval(g, XYZ))) then return false; endif;
  endforeach;
  return true;
enddefine; -- IsSolution

foreach XYZ in CartesianProductList([Xs,Ys,Zs]) do
  if IsSolution(XYZ, I) then println "yes :-) ", XYZ;
  else println "-- no :-( ", XYZ;  --> emacs prints it red!
  endif;
endforeach;

--------------------------------------------------
-- THE "GOOD" ALGORITHM

-- input: the "polynomial system" I
S := [I];

----------
X := x; -- fix the first indeterminate

J := I;  -- S contains just the ideal I

EF := eigenfactors(MultiplicationMat(X, J),  X);
println "----> EF = ", EF;
LinFac := [ ideal(l) | l in EF and deg(l)=1 ]; -- linear factors
println "----> LinFac = ", LinFac;
NewS := [ ideal(ReducedGBasis(J + L)) | L in LinFac ];

S := NewS;  indent(S);

----------
X := y; -- fix the second indeterminate
NewS := [];
foreach J in S do
  println "----> J = ", J;
  EF := eigenfactors(MultiplicationMat(X, J),  X);
  println "----> EF = ", EF;
  LinFac := [ ideal(l) | l in EF and deg(l)=1 ]; -- linear factors
  println "----> LinFac = ", LinFac;
  NewS := concat(NewS, [ ideal(ReducedGBasis(J + L)) | L in LinFac ]);
endforeach;
S := NewS;  indent(S);

----------
X := z; -- fix the third indeterminate
NewS := [];
foreach J in S do
  println "----> J = ", J;
  EF := eigenfactors(MultiplicationMat(X, J),  X);
  println "----> EF = ", EF;
  LinFac := [ ideal(l) | l in EF and deg(l)=1 ]; -- linear factors
  println "----> LinFac = ", LinFac;
  NewS := concat(NewS, [ ideal(ReducedGBasis(J + L)) | L in LinFac ]);
endforeach;
S := NewS;  indent(S);


foreach J in S do
  println NF(x,J), " ", NF(y,J), " ", NF(z,J);
endforeach;

RationalSolve(gens(I));

-- EXERCISE: try with other 0-dimensional ideals
-- EXERCISE: write the function RationalSolve_tut

-- ------------------------------------------------------------
-- -- SPLITTING A POLYNOMIAL
-- ------------------------------------------------------------
-- FFp := NewZZmod(101);
-- use Kt ::= FFp[t];
-- f := t^2 +3;
-- IsIrred(f); --> true

-- L := Kt/ideal(f);
-- Use L;

-- p := characteristic(L);
-- e := LogCardinality(L);

-- Use L[x];
-- x^2 +3 = product([ x-t^(p^i) | i in 0..e-1]);
-- --> L is a splitting field of f(x)

-- ------------------------------------------------------------
-- -- ALGORITHM (COMPUTING SOLUTIONS OVER FINITE FIELDS)
-- ------------------------------------------------------------

-- Use L[x,y,z];
-- M := IntersectList([ideal(z^2-4, y^3-8, (z-x)*(x-3)),
-- 		    ideal(x^2-1, y^2-1, x+y+z)
-- 		    ]);
-- indent(I);
-- RingOf(I);
-- describe RingOf(I);
-- -- PROBLEM. Find the L-rational solutions of I:

-- Ktx ::= FFp[t, x,y,z];
-- -- read "I" here:  (no homomorphism L[x,y,z]-->FFp[t, x,y,z]!)
-- -- one should make a proper mathematical construction,
-- -- but there is a trick: (string)print it and read it in Ktx
-- G := gens(I);
-- Gtx := [ ReadExpr(Ktx, sprint(g)) | g in G ];
-- indent(Gtx);

-- -- EXERCISE: what it is the proper mathematical construction?

-- J := ideal(Gtx) + ideal(ReadExpr(Ktx, sprint(f)));
-- RingOf(J);
-- describe RingOf(J);
-- PrDec := PrimaryDecomposition0(J);  indent(PrDec, 2);

-- lascerei perdere....

----------------------------------------------------------------------
-- APPROXIMATION

-- EXAMPLE 1

use P;
f1 := *** 3x^2 -y^2 +2yz -z^2 -8x -8y +5z-5 ***;
f2 := *** x^3 -6x^2 -6xy-4y^2  +z^2  +3x +7y-7z +15 ***;
f3 := *** z^3  +4x^2 +2xy-3z^2 -13x-5y +6z +5 ***;

I := ideal(f1, f2, f3);

-- input: the "polynomial system" I
S := [I];

----------
X := x; -- fix the first indeterminate

J := I;  -- S contains just the ideal I

EF := eigenfactors(MultiplicationMat(X, J),  X);
print "----> EF = "; indent(EF);
NewS := [ ideal(ReducedGBasis(J + ideal(f))) | f in EF ];

S := NewS;  indent(S);
-->   ideal(z -1, y +1, x -2) --> rational solution


use Plex ::= QQ[x,y,z], lex;
phi := PolyAlgebraHom(P, Plex, indets(Plex));

J := ideal(phi(gens(S[1])));
RGB := ReducedGBasis(J);  indent(RGB);
--> shape lemma: linear in x and y, and z^15 ...

gz := last(RGB);
IsIrred(gz);  --> J is maximal in QQ[x,y,z]
--> this is the best we can do with exact computations

-- APPROXIMATION:
RR := RealRootsApprox(gz, 10^(-30));
indent(RR);
--> three real roots for gz = 0

indent([DecimalStr(r, 30) | r in RR]);

foreach r in RR do
  Jr := ideal(diff(RGB, [gz])) + ideal(z-r);
  Sol := [NF(x,Jr), NF(y,Jr), NF(z,Jr)];
  println [DecimalStr(c) | c in Sol];
endforeach;

-- EXAMPLE 2

I := IntersectList([ideal(z^2-y, y^2-8, (y-x)*(x-3)),
		    ideal(x^2-3, y^3-1, x+y+z)
		    ]);
indent(I);

-- the "good" algorithm for all real solutions:
--------------------------------------------------
-- input: the "polynomial system" I
S := [I];

----------
X := x; -- fix the first indeterminate

J := I;  -- S contains just the ideal I

EF := eigenfactors(MultiplicationMat(X, J),  X);
println "----> EF = ", EF;
RealFac := [ ideal(f) | f in EF and RealRootsApprox(f)<>[] ];
println "----> RealFac = ", RealFac;
NewS := [ ideal(ReducedGBasis(J + L)) | L in RealFac ];

S := NewS;  indent(S);

----------
X := y; -- fix the second indeterminate
NewS := [];
foreach J in S do
  println "----> J = ", J;
  EF := eigenfactors(MultiplicationMat(X, J),  X);
  println "----> EF = ", EF;
  RealFac := [ ideal(f) | f in EF and RealRootsApprox(f)<>[] ];
  println "----> RealFac = ", RealFac;
  NewS := concat(NewS, [ ideal(ReducedGBasis(J + L)) | L in RealFac ]);
endforeach;
S := NewS;  indent(S);

----------
X := z; -- fix the third indeterminate
NewS := [];
foreach J in S do
  println "----> J = ", J;
  EF := eigenfactors(MultiplicationMat(X, J),  X);
  println "----> EF = ", EF;
  RealFac := [ ideal(f) | f in EF and RealRootsApprox(f)<>[] ];
  println "----> RealFac = ", RealFac;
  NewS := concat(NewS, [ ideal(ReducedGBasis(J + L)) | L in RealFac ]);
endforeach;
S := NewS;  indent(S);

-- and now....
-- EXERCISE: write all rational solutions
-- EXERCISE: write all (approximate) real solutions

-- EXERCISE: try with other 0-dimensional ideals

-- EXERCISE: experiment with all what you have learned!!

--   enjoy CoCoA!
