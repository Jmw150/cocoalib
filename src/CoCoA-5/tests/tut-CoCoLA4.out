ideal(
  x^2*y +(1/2)*x^2*z -x*y*z +(3/4)*y*z^2 -3*x*y +(-3/2)*x*z +3*y*z +(9/4)*z^2 -2*x -3*y -3,
  y^3 +(23/48)*x^2*z +(55/96)*y*z^2 +(-5/24)*x^2 +(-59/48)*x*z +(-59/96)*z^2 +(-31/24)*x +(-55/24)*y +(-5/8)*z +5/24,
  z^3 -4*z,
  y^2*z^2 -4*y^2 -z^2 +4,
  x^3 +(-1/2)*x^2*z +(1/4)*y*z^2 -3*x^2 +(3/2)*x*z +(3/4)*z^2 -2*x -y +3,
  x*z^2 +y*z^2 -4*x -4*y
)
[-1,  1,  -2,  2,  3]
[-1,  1,  2]
[0,  -2,  2]
-- no :-( [-1,  -1,  0]
-- no :-( [-1,  -1,  -2]
yes :-) [-1,  -1,  2]
yes :-) [-1,  1,  0]
-- no :-( [-1,  1,  -2]
-- no :-( [-1,  1,  2]
-- no :-( [-1,  2,  0]
-- no :-( [-1,  2,  -2]
-- no :-( [-1,  2,  2]
yes :-) [1,  -1,  0]
-- no :-( [1,  -1,  -2]
-- no :-( [1,  -1,  2]
-- no :-( [1,  1,  0]
yes :-) [1,  1,  -2]
-- no :-( [1,  1,  2]
-- no :-( [1,  2,  0]
-- no :-( [1,  2,  -2]
-- no :-( [1,  2,  2]
-- no :-( [-2,  -1,  0]
-- no :-( [-2,  -1,  -2]
-- no :-( [-2,  -1,  2]
-- no :-( [-2,  1,  0]
-- no :-( [-2,  1,  -2]
-- no :-( [-2,  1,  2]
-- no :-( [-2,  2,  0]
yes :-) [-2,  2,  -2]
-- no :-( [-2,  2,  2]
-- no :-( [2,  -1,  0]
-- no :-( [2,  -1,  -2]
-- no :-( [2,  -1,  2]
-- no :-( [2,  1,  0]
-- no :-( [2,  1,  -2]
-- no :-( [2,  1,  2]
-- no :-( [2,  2,  0]
-- no :-( [2,  2,  -2]
yes :-) [2,  2,  2]
-- no :-( [3,  -1,  0]
-- no :-( [3,  -1,  -2]
-- no :-( [3,  -1,  2]
-- no :-( [3,  1,  0]
-- no :-( [3,  1,  -2]
-- no :-( [3,  1,  2]
-- no :-( [3,  2,  0]
yes :-) [3,  2,  -2]
yes :-) [3,  2,  2]
----> EF = [x +1,  x -1,  x +2,  x -2,  x -3]
----> LinFac = [ideal(x +1),  ideal(x -1),  ideal(x +2),  ideal(x -2),  ideal(x -3)]
[
  ideal(x +1,  z^2 -2*z,  y +z -1),
  ideal(x -1,  z^2 +2*z,  y +z +1),
  ideal(x +2,  y^3 -8,  z +2),
  ideal(x -2,  y^3 -8,  z -2),
  ideal(x -3,  y^3 -8,  z^2 -4)
]
----> J = ideal(x +1,  z^2 -2*z,  y +z -1)
----> EF = [y +1,  y -1]
----> LinFac = [ideal(y +1),  ideal(y -1)]
----> J = ideal(x -1,  z^2 +2*z,  y +z +1)
----> EF = [y +1,  y -1]
----> LinFac = [ideal(y +1),  ideal(y -1)]
----> J = ideal(x +2,  y^3 -8,  z +2)
----> EF = [y -2,  y^2 +2*y +4]
----> LinFac = [ideal(y -2)]
----> J = ideal(x -2,  y^3 -8,  z -2)
----> EF = [y -2,  y^2 +2*y +4]
----> LinFac = [ideal(y -2)]
----> J = ideal(x -3,  y^3 -8,  z^2 -4)
----> EF = [y -2,  y^2 +2*y +4]
----> LinFac = [ideal(y -2)]
[
  ideal(y +1,  z -2,  x +1),
  ideal(y -1,  z,  x +1),
  ideal(y +1,  z,  x -1),
  ideal(y -1,  z +2,  x -1),
  ideal(z +2,  y -2,  x +2),
  ideal(z -2,  y -2,  x -2),
  ideal(y -2,  x -3,  z^2 -4)
]
----> J = ideal(y +1,  z -2,  x +1)
----> EF = [z -2]
----> LinFac = [ideal(z -2)]
----> J = ideal(y -1,  z,  x +1)
----> EF = [z]
----> LinFac = [ideal(z)]
----> J = ideal(y +1,  z,  x -1)
----> EF = [z]
----> LinFac = [ideal(z)]
----> J = ideal(y -1,  z +2,  x -1)
----> EF = [z +2]
----> LinFac = [ideal(z +2)]
----> J = ideal(z +2,  y -2,  x +2)
----> EF = [z +2]
----> LinFac = [ideal(z +2)]
----> J = ideal(z -2,  y -2,  x -2)
----> EF = [z -2]
----> LinFac = [ideal(z -2)]
----> J = ideal(y -2,  x -3,  z^2 -4)
----> EF = [z +2,  z -2]
----> LinFac = [ideal(z +2),  ideal(z -2)]
[
  ideal(z -2,  y +1,  x +1),
  ideal(z,  y -1,  x +1),
  ideal(z,  y +1,  x -1),
  ideal(z +2,  y -1,  x -1),
  ideal(z +2,  y -2,  x +2),
  ideal(z -2,  y -2,  x -2),
  ideal(z +2,  y -2,  x -3),
  ideal(z -2,  y -2,  x -3)
]
-1 -1 2
-1 1 0
1 -1 0
1 1 -2
-2 2 -2
2 2 2
3 2 -2
3 2 2
record[AffinePts := [[-2,  2,  -2],  [-1,  -1,  2],  [-1,  1,  0],  [1,  -1,  0],  [1,  1,  -2],  [2,  2,  2],  [3,  2,  -2],  [3,  2,  2]], indets := [x,  y,  z]]
----> EF = [
  x^15 -102*x^14 +4254*x^13 -91420*x^12 +1108030*x^11 -7617460*x^10 +26368737*x^9 -16709678*x^8 -163401402*x^7 +408143116*x^6 -102334132*x^5 -561462068*x^4 +119765981*x^3 +1225188788*x^2 -1326547350*x +524273780,
  x -2
]
[
  ideal(x^2 +(-1/3)*y^2 +(2/3)*y*z +(-1/3)*z^2 +(-8/3)*x +(-8/3)*y +(5/3)*z -5/3,  z^3 +2*x*y +(4/3)*y^2 +(-8/3)*y*z +(-5/3)*z^2 +(-7/3)*x +(17/3)*y +(-2/3)*z +35/3,  x*y^2 -2*x*y*z +x*z^2 -10*x*y +(-46/3)*y^2 -5*x*z +(20/3)*y*z +(-1/3)*z^2 +(-38/3)*x +(-17/3)*y +(-13/3)*z +85/3,  y^4 -10*y^2*z^2 +(-82/3)*y^3 -282*x*y*z +50*y^2*z +103*x*z^2 +4*y*z^2 +(-2917/3)*x*y +(-13069/9)*y^2 +(-1438/3)*x*z +(5033/9)*y*z +(845/9)*z^2 +(-13034/9)*x +(-10751/9)*y +(-3754/9)*z +20215/9,  y^3*z -4*y^2*z^2 +(-23/3)*y^3 -56*x*y*z +17*y^2*z +(33/2)*x*z^2 +(-5/2)*y*z^2 +(-437/3)*x*y +(-2111/9)*y^2 +(-209/3)*x*z +(820/9)*y*z +(449/18)*z^2 +(-2323/9)*x +(-2176/9)*y +(-1387/18)*z +3065/9,  x*y*z^2 +(1/3)*y^2*z^2 +(20/9)*y^3 +(-56/9)*y^2*z +(-83/3)*x*z^2 +(-122/9)*y*z^2 +(280/3)*x*y +(2434/27)*y^2 +(124/9)*x*z +(-893/27)*y*z +(295/27)*z^2 +(3002/27)*x +(320/27)*y +(421/27)*z -4030/27),
  ideal(x -2,  y +1,  z -1)
]
[
  x +(74220620119139098747697747088683505655/67466931478152492327163393797513474054999076)*z^14 +(-265457712060172581958838044351678739618/16866732869538123081790848449378368513749769)*z^13 +(16443100207920016822953002412557360043449/33733465739076246163581696898756737027499538)*z^12 +(-105095553509273436406508704266501011149737/67466931478152492327163393797513474054999076)*z^11 +(12583591060121599600383895400533466246285/16866732869538123081790848449378368513749769)*z^10 +(2987499874568925020025898227945895399521557/67466931478152492327163393797513474054999076)*z^9 +(-8575141569995136553775235643904126485826767/67466931478152492327163393797513474054999076)*z^8 +(-1366740790354950372640183504711890666071323/33733465739076246163581696898756737027499538)*z^7 +(108566501393988418342836640539275359812341499/67466931478152492327163393797513474054999076)*z^6 +(-92930952508332917868912718636713388112001837/33733465739076246163581696898756737027499538)*z^5 +(1317218870412200311581017801052672293173997/67466931478152492327163393797513474054999076)*z^4 +(920548037484359195555903206827271340763441969/67466931478152492327163393797513474054999076)*z^3 +(-1049991567654819272068408226185792083280896789/67466931478152492327163393797513474054999076)*z^2 +(902001516539004486836254756443788891441146599/33733465739076246163581696898756737027499538)*z -430840396754847917953148988229397289956555830/16866732869538123081790848449378368513749769,
  y +(-158806623619616673990040830604606786353/67466931478152492327163393797513474054999076)*z^14 +(1132752290911997482099028205887507436469/33733465739076246163581696898756737027499538)*z^13 +(-35138147812149131928854566121181642580285/33733465739076246163581696898756737027499538)*z^12 +(222038365864017714803851146583975971888059/67466931478152492327163393797513474054999076)*z^11 +(-49845501347470117992101095863644657061439/33733465739076246163581696898756737027499538)*z^10 +(-6402371334220401096522710644996676571500729/67466931478152492327163393797513474054999076)*z^9 +(18150700528639151598337078648952624320488267/67466931478152492327163393797513474054999076)*z^8 +(3131571573805051913279737666293497477022717/33733465739076246163581696898756737027499538)*z^7 +(-232123995771734638592493533888111250050558389/67466931478152492327163393797513474054999076)*z^6 +(98109098105670430996478678337004454105832323/16866732869538123081790848449378368513749769)*z^5 +(-2597027754255495164719692919444801536360255/67466931478152492327163393797513474054999076)*z^4 +(-1969243797772608514575470991340819294452913633/67466931478152492327163393797513474054999076)*z^3 +(2232904877524345107473226725603143436420865041/67466931478152492327163393797513474054999076)*z^2 +(-1984948260809364577137541278020621522545726269/33733465739076246163581696898756737027499538)*z +896858170780611809746689830945475150465533475/16866732869538123081790848449378368513749769,
  z^15 -15*z^14 +453*z^13 -1723*z^12 +1653*z^11 +39878*z^10 -143949*z^9 +44358*z^8 +1493331*z^7 -3545819*z^6 +1822878*z^5 +12465182*z^4 -23275915*z^3 +35305982*z^2 -40170300*z +15978680
]
true
[
  -1711547769131389396763452378925159814953/680564733841876926926749214863536422912,
  492564573041320396882943614343070389753/680564733841876926926749214863536422912,
  10869469019501478906626664660822267954079/10889035741470030830827987437816582766592
]
[
  "-2.514893417220546241742308949488",
  "0.723758591281580172944362357520",
  "0.998203080379832567529146806288"
]
["-2.373",  "0.862",  "-2.515"]
["9.492",  "-17.531",  "0.724"]
["2.030",  "-0.970",  "0.998"]
ideal(
  x^2 -x*z^2 -3*x +(42646431/1008664540)*z^9 +(28155506/756498405)*z^8 +(-183439063/605198724)*z^7 +(-4345053/38794790)*z^6 +(264598681/605198724)*z^5 +(345667489/504332270)*z^4 +(366878126/151299681)*z^3 +(75572397/19397395)*z^2 +(-4693015498/756498405)*z -457689404/58192185,
  y +(-2366684/756498405)*z^9 +(7499327/3025993620)*z^8 +(676483/100866454)*z^7 +(-7896209/232768740)*z^6 +(3656811/50433227)*z^5 +(63795947/3025993620)*z^4 +(-2705932/50433227)*z^3 +(-42399767/58192185)*z^2 +(-287349544/756498405)*z -6348234/19397395,
  x*z^4 -8*x +(8/195)*z^9 +(-1/195)*z^8 +(-16/39)*z^7 +(2/15)*z^6 +(59/39)*z^5 +(33/65)*z^4 +(128/39)*z^3 +(-16/15)*z^2 +(-2872/195)*z -56/15,
  z^10 -9*z^8 +2*z^7 +19*z^6 +18*z^5 +46*z^4 -16*z^3 -216*z^2 -144*z +208
)
----> EF = [x^2 -8,  x^2 -3,  x -3]
----> RealFac = [ideal(x^2 -8),  ideal(x^2 -3),  ideal(x -3)]
[
  ideal(y -z^2,  x -z^2,  z^4 -8),
  ideal(y +(-8/195)*z^5 +(1/195)*z^4 +(16/39)*z^3 +(-2/15)*z^2 +(-164/195)*z -7/15,  x +(8/195)*z^5 +(-1/195)*z^4 +(-16/39)*z^3 +(2/15)*z^2 +(359/195)*z +7/15,  z^6 -9*z^4 +2*z^3 +27*z^2 +18*z -26),
  ideal(y -z^2,  x -3,  z^4 -8)
]
----> J = ideal(y -z^2,  x -z^2,  z^4 -8)
----> EF = [y^2 -8]
----> RealFac = [ideal(y^2 -8)]
----> J = ideal(y +(-8/195)*z^5 +(1/195)*z^4 +(16/39)*z^3 +(-2/15)*z^2 +(-164/195)*z -7/15,  x +(8/195)*z^5 +(-1/195)*z^4 +(-16/39)*z^3 +(2/15)*z^2 +(359/195)*z +7/15,  z^6 -9*z^4 +2*z^3 +27*z^2 +18*z -26)
----> EF = [y -1,  y^2 +y +1]
----> RealFac = [ideal(y -1)]
----> J = ideal(y -z^2,  x -3,  z^4 -8)
----> EF = [y^2 -8]
----> RealFac = [ideal(y^2 -8)]
[
  ideal(y -z^2,  x -z^2,  z^4 -8),
  ideal(y -1,  x +z +1,  z^2 +2*z -2),
  ideal(y -z^2,  x -3,  z^4 -8)
]
----> J = ideal(y -z^2,  x -z^2,  z^4 -8)
----> EF = [z^4 -8]
----> RealFac = [ideal(z^4 -8)]
----> J = ideal(y -1,  x +z +1,  z^2 +2*z -2)
----> EF = [z^2 +2*z -2]
----> RealFac = [ideal(z^2 +2*z -2)]
----> J = ideal(y -z^2,  x -3,  z^4 -8)
----> EF = [z^4 -8]
----> RealFac = [ideal(z^4 -8)]
[
  ideal(y -z^2,  x -z^2,  z^4 -8),
  ideal(y -1,  x +z +1,  z^2 +2*z -2),
  ideal(y -z^2,  x -3,  z^4 -8)
]
